### DISCUSSION TITLE
How are fonts handled on MCU exactly?

### DISCUSSION THREAD

**Discussion by yanshay** [slint-ui/slint#6302] [Q&A]
Is there documentation of how fonts are handled on MCU with Software Renderer with all details?
I've seen some information spread in several places but these pieces don't cover everything.

A few questions for examples that I have:
1. What font is picked to be included as default font? Is it OS dependent? So if I compile on different OS will the app ilook exactly the same with the default font?
2. Which font sizes are included and how can I control the sizes that will be included if the code sets dynamically the font size?
3. Same goes for font weight and other properties.
4. Which characters are included automatically? I saw when it comes to foreign languages not all are included automatically?
5. Symbols - some symbols I was able to display and some I didn't. That might depend on the font picked but I had cases of symbols I couldn't display that seemed awkward not to be included in a font if others were.
6. Is it possible to use emoji fonts? I could't get these to work on MCU (when it worked on the live preview) - they are multicolored.

**Comment by ogoffart** [comment]
On MCU, fonts are embedded in the binary by the slint compiler

> 1. What font is picked to be included as default font? Is it OS dependent? So if I compile on different OS will the app ilook exactly the same with the default font?

I think it might indeed be different as it pick the default font from the system.
It can be usefull to import font and use that one to be sure to use a specific font.

> 2. Which font sizes are included and how can I control the sizes that will be included if the code sets dynamically the font size?

It tries some heuristics by looking at the size of the Text element at compile time
It can also be configured by an env variable while compiling
eg: `SLINT_FONT_SIZES=12,16,24`

> Which characters are included automatically? I saw when it comes to foreign languages not all are included automatically?

It look at all string literal in the file and encode all the character seen. In addition to ascii and a few internal ones:
https://github.com/slint-ui/slint/blob/cfa115affafc48293a545860d5368de28c7e6467/internal/compiler/passes/embed_glyphs.rs#L52-l58

> 5. Symbols - some symbols I was able to display and some I didn't. That might depend on the font picked but I had cases of symbols I couldn't display that seemed awkward not to be included in a font if others were.

See previous question. if the compiler sees it, it might integrate it.

> 6. Is it possible to use emoji fonts? I could't get these to work on MCU (when it worked on the live preview) - they are multicolored.

Also color emoji are not yet supported.

**Reply by yanshay** [reply]
Thanks, that's very helpful.

> I think it might indeed be different as it pick the default font from the system.
> It can be usefull to import font and use that one to be sure to use a specific font.

If I import a font, I need to explicitly ask for that font family, right?
And then will it import also the system font or would it be instead? (to save on binary size)

> SLINT_FONT_SIZES=12,16,24

Is there something like that for FONT_WEIGHTS ?


Worth adding in the fonts doc. Now that I searched for it I see it's mentioned in several discussions but not in the documentation of font.
